frame/cosmwasm/src/lib.rs

```diff
@@ -0,0 +1,1071 @@
+			message: ContractMessageOf<T>,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+			let mut shared = Self::do_create_vm_shared(gas, false);
+			let outcome = Self::do_extrinsic_execute(&mut shared, who, contract, funds, message);
+			Self::refund_gas(outcome, gas, shared.gas.remaining())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Deterministic contract address computation, equivalent to solidity CREATE2.
+		pub(crate) fn derive_contract_address(
+			instantiator: &AccountIdOf<T>,
+			code_id: CosmwasmCodeId,
+			salt: &[u8],
+		) -> AccountIdOf<T> {
+			let data: Vec<_> = instantiator
+				.as_ref()
+				.iter()
+				.chain(&code_id.to_le_bytes())
+				.chain(salt)
+				.cloned()
+				.collect();
+			UncheckedFrom::unchecked_from(T::Hashing::hash(&data))
+		}
+
+		/// Deterministic contract trie id generation.
+		pub(crate) fn derive_contract_trie_id(
+			contract: &AccountIdOf<T>,
+			nonce: u64,
+		) -> ContractTrieIdOf<T> {
+			let data: Vec<_> =
+				contract.as_ref().iter().chain(&nonce.to_le_bytes()).cloned().collect();
+			T::Hashing::hash(&data).as_ref().to_vec().try_into().expect(
+				"hashing len implementation must always be <= defined max contract trie id size; QED;",
+			)
+		}
+
+		/// Setup pallet state for a new contract.
+		/// This is called prior to calling the `instantiate` export of the contract.
+		pub(crate) fn do_instantiate_phase1(
+			instantiator: AccountIdOf<T>,
+			code_id: CosmwasmCodeId,
+			salt: &[u8],
+			admin: Option<AccountIdOf<T>>,
+			label: ContractLabelOf<T>,
+		) -> Result<(AccountIdOf<T>, ContractInfoOf<T>), Error<T>> {
+			let contract = Self::derive_contract_address(&instantiator, code_id, salt);
+			ensure!(
+				!ContractToInfo::<T>::contains_key(&contract),
+				Error::<T>::ContractAlreadyExists
+			);
+			let nonce =
+				CurrentNonce::<T>::increment().expect("2305843009 contracts per human; qed;");
+			let trie_id = Self::derive_contract_trie_id(&contract, nonce);
+			let contract_info =
+				ContractInfoOf::<T> { instantiator, code_id, trie_id, admin, label };
+			ContractToInfo::<T>::insert(&contract, &contract_info);
+			CodeIdToInfo::<T>::try_mutate(code_id, |entry| -> Result<(), Error<T>> {
+				let code_info = entry.as_mut().ok_or(Error::<T>::CodeNotFound)?;
+				code_info.refcount += 1;
+				Ok(())
+			})?;
+			Self::deposit_event(Event::<T>::Instantiated {
+				contract: contract.clone(),
+				info: contract_info.clone(),
+			});
+			Ok((contract, contract_info))
+		}
+
+		/// Create the VM shared state. Including readonly stack, VM depth, gas metering limits and
+		/// code cache.
+		///
+		/// This state is shared accross all VMs (all contracts loaded within a single call) and is
+		/// used to optimize some operations as well as track shared state (readonly storage while
+		/// doing a `query` etc...)
+		pub(crate) fn do_create_vm_shared(gas: u64, readonly: bool) -> CosmwasmVMShared {
+			CosmwasmVMShared {
+				readonly: if readonly { 1 } else { 0 },
```

Comment:
Should we change the definition of `CosmwasmVMShared` do have an enum field, and go for something like:

```
do_create_vm_shared(100, Mutability::ReadOnly)
vs
do_create_vm_shared(100, true)
```

Reply:
The thing is the `CosmwasmVMSharted` struct contains a `readonly` counter (not a boolean).
I.e. if you have contract `A` and `B`, call `query` on `A`, the `readonly` counter is incremented/decremented before/after the `query` call. This avoid a situation where you  `query(A)` and within the query, `A` do a `query(B)` => 2 readonly depth, we can't set readonly = false when we exit `query(B)` or `query(A)` would be able to mutate the state after the call to `query(B)`.

Reply:
I'll still introduce an enum for the initial counter increment

Reply:
Or just the function definitions. If the counter makes sense inside the VM, let's keep it at that.

Reply:
Yep!